### PR TITLE
[spritesmith] Add types

### PR DESCRIPTION
**Pull request by Zenoo**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Select one of these and delete the others:

If adding a new definition:
- [x] The package does not already provide its own types, or cannot have its `.d.ts` files generated via `--declaration`
- [x] If this is for an npm package, match the name. If not, do not conflict with the name of an npm package.
- [x] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [x] Represents shape of module/library [correctly](https://www.typescriptlang.org/docs/handbook/declaration-files/library-structures.html)
- [x] `tslint.json` [should contain](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#linter-tslintjson) `{ "extends": "@definitelytyped/dtslint/dt.json" }`, and no additional rules.
- [x] `tsconfig.json` [should have](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#tsconfigjson) `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`.

### CODE REVIEW

File: types/spritesmith/index.d.ts [hunk 1]
```diff
@@ -0,0 +1,60 @@
+// Type definitions for spritesmith 3.4
+// Project: https://github.com/twolfson/spritesmith
+// Definitions by: Zenoo <https://github.com/Zenoo>
+// Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
+
+import { ReadableStream } from 'stream/web';
+import { BufferFile } from 'vinyl';
+
+interface SpritesmithOptions {
```
Comment (peterblazejewicz) [inline types/spritesmith/index.d.ts:9]: they call this 'Params' for some reason and are using all other the readme:
https://github.com/twolfson/spritesmith#new-spritesmithparams

Comment (Zenoo) [reply types/spritesmith/index.d.ts]: I changed it to SpritesmithParams, although I don't really see why that's important ^^

Comment (peterblazejewicz) [reply types/spritesmith/index.d.ts]: just sticking with original intent/naming, the name does not really matter in TS, if shape of type is same

File: types/spritesmith/tsconfig.json [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+    "compilerOptions": {
+        "module": "commonjs",
+        "lib": [
+            "es6"
+        ],
+        "noImplicitAny": true,
+        "noImplicitThis": true,
+        "strictFunctionTypes": true,
+        "strictNullChecks": true,
+        "esModuleInterop": true,
```
Comment (peterblazejewicz) [inline types/spritesmith/tsconfig.json:11]:
```suggestion
        "strictNullChecks": true,
```
https://github.com/DefinitelyTyped/DefinitelyTyped#esmoduleinteropallowsyntheticdefaultimports
that's meant to be used when consuming TS libraries/packages, not recommended here

File: types/spritesmith/spritesmith-tests.ts [hunk 3]
```diff
@@ -0,0 +1,25 @@
+import Spritesmith from "spritesmith";
```
Comment (peterblazejewicz) [inline types/spritesmith/spritesmith-tests.ts:1]:
```suggestion
import Spritesmith = require("spritesmith");
```

Comment (peterblazejewicz) [reply types/spritesmith/spritesmith-tests.ts:1]: (ditto, see comment about es module imports)

**Review by peterblazejewicz — CHANGES_REQUESTED** [review]
please re-review comments about `esModuleInterop`, thanks!

**Comment by Zenoo** [comment]
@peterblazejewicz I can't remove this tag as it produces errors while testing:

**Comment by peterblazejewicz** [comment]
> @peterblazejewicz I can't remove this tag as it produces errors while testing: 

Please review my comments, marked as resolved. You have to update import in test to use `import ... = require(....);` syntax. thanks!

**Comment by Zenoo** [comment]
Oh, weird, I accepted the review modifications, but the commits seemingly disappeared, sorry about that, it should be fixed now.

**Review by peterblazejewicz — APPROVED** [review]

**Review by peterblazejewicz — APPROVED** [review]
LGTM!
@Zenoo thanks!

**Comment by Zenoo** [comment]
Ready to merge

**Comment by Zenoo** [comment]
@peterblazejewicz I tested the `@types/spritesmith` package and this reviewed version does not allow for people to use the declared types, like it could in my first version.  

I can't import the type SpritesmithResult, for example.  

Is there any way to change those declarations to allow imports?

**Comment by peterblazejewicz** [comment]
everything depends on specific setup. Can you try this ne with default TS project configuration?
```ts
import Spritesmith = require('spritesmith');

const sprites = ['fork.png', 'github.png', 'twitter.png'];
Spritesmith.run({ src: sprites }, function handleResult(err, result) {
    result.image;
    result.coordinates;
    result.properties;
});
```

**Comment by peterblazejewicz** [comment]
If one has default (for new TS configs) `esModuleInterop` option enabled, this works:
```ts
import Spritesmith from 'spritesmith';

const sprites = ['fork.png', 'github.png', 'twitter.png'];
Spritesmith.run({ src: sprites }, function handleResult(err, result) {
    result.image;
    result.coordinates;
    result.properties;
});
```
producing:
```js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const spritesmith_1 = __importDefault(require("spritesmith"));
const sprites = ['fork.png', 'github.png', 'twitter.png'];
spritesmith_1.default.run({ src: sprites }, function handleResult(err, result) {
    result.image;
    result.coordinates;
    result.properties;
});
```

**Comment by peterblazejewicz** [comment]
image

**Comment by Zenoo** [comment]
No, I mean, people need to be able to import the every type declared here for use elsewhere.  
If for some reason I need to give the type `SpritesmithResult` to some variable like:
```ts
let result: SpritesmithResult | null = null;
```

It's not possible, since there is no way to import the type `SpritesmithResult` from the declaration

**Comment by peterblazejewicz** [comment]
here you are, put your types into 'namespace' to export, rewritten:
```ts
// Type definitions for spritesmith 3.4
// Project: https://github.com/twolfson/spritesmith
// Definitions by: Zenoo <https://github.com/Zenoo>
// Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped

import { ReadableStream } from 'stream/web';
import { BufferFile } from 'vinyl';

declare class Spritesmith {
    constructor(params?: Spritesmith.SpritesmithParams);
    static run(
        params: Spritesmith.SpritesmithParams & Spritesmith.SpritesmithProcessImagesOptions & {
            src: Spritesmith.SpritesmithCreateImagesSrc;
        },
        callback: (
            err: Error | null,
            result: Spritesmith.SpritesmithResult & {
                image: Buffer;
            }
        ) => void): void;
    createImages(src: Spritesmith.SpritesmithCreateImagesSrc, callback: (err: Error | null, images: SpritesmithImage[]) => void): void;
    processImages(
        images: Spritesmith.SpritesmithImage[],
        options?: Spritesmith.SpritesmithProcessImagesOptions
    ): Spritesmith.SpritesmithResult;
}

declare namespace Spritesmith {
    interface SpritesmithParams {
        engine?: string;
        engineOpts?: Record<string, unknown>;
    }

    type SpritesmithCreateImagesSrc = Array<string | BufferFile>;
    interface SpritesmithImage {
        width: number;
        height: number;
        [key: string]: unknown;
    }

    interface SpritesmithProcessImagesOptions {
        padding?: number;
        exportOpts?: {
            format?: 'png' | 'jpg' | 'jpeg' | 'webp';
            quality?: number;
            background?: string;
            [key: string]: unknown;
        };
        algorithm?: 'top-down' | 'left-right' | 'diagonal' | 'alt-diagonal' | 'binary-tree';
        algorithmOpts?: {
            sort?: boolean;
        };
    }

    interface SpritesmithResult {
        image: ReadableStream;
        coordinates: Record<string, { x: number; y: number; width: number; height: number }>;
        properties: { width: number; height: number };
    }
}

export = Spritesmith;

````

usage, e.g.:
```ts
import Spritesmith, {
    SpritesmithCreateImagesSrc,
    SpritesmithImage,
    SpritesmithParams,
    SpritesmithProcessImagesOptions,
    SpritesmithResult
} from 'spritesmith';
```

**Comment by Zenoo** [comment]
Thanks ! I will open a PR with these modifications

**Pull request merged**